### PR TITLE
Using `fetch_min` / `fetch_max` in `__parallel_find_or` and in it's stuff

### PR DESCRIPTION
**Pull request by SergeyKopienko**
In this PR we using `fetch_min` / `fetch_max` in `__parallel_find_or` and it's stuff.
This approach give us additional performance boost:
- now we shouldn't try to setup atomic states inside `for`-based loops with additional checks.

This idea was proposed by @danhoeflinger.

Also in this PR we using `__dpl_ceiling_div` function where it's applicable in the code.

### CODE REVIEW

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (julianmi) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1155]: `sizeof` uses bytes so this should be `sizeof(_AtomicType) < 8`. I think even better would be `<= 4` since only 32-bit data types are then allowed.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (julianmi) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1041]: The processing order isn't relevant here. We might want to simplify this to something like: "Store that a match was found. Its position is not relevant for or semantics."

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (julianmi) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1233]: This comes probably down to preference but I preferred the approach before. I think it was easier to read and didn't require duplicated routines in forward/backward tag.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (danhoeflinger) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1154]: This is a bigger question than this PR is looking to address, but should we have a "general" version of this algorithm which supports hardware which doesn't support 64-bit atomics rather than merely an assert? Obviously this would come at a cost, and additional code.

For instance, I believe this is not supported in FPGA (if I remember correctly from work on histogram), which also may use 64-bit atomics based on the types passed in from the user).

Note that this is not a new problem added by this PR or any recent work, but just one which is been brought up by the addition of the assertion.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1155]: Good catch, @julianmi 
But this mean we have the same error here:
```C++
    assert("This device does not support 64-bit atomics" &&
           (sizeof(_Type) < 64 || __queue.get_device().has(sycl::aspect::atomic64)));
```
I mean https://github.com/oneapi-src/oneDPL/blob/e1fee6404e4ebda93ade0d3e913e1eb9e9b8595b/include/oneapi/dpl/experimental/kt/single_pass_scan.h#L328
It was introduced in https://github.com/oneapi-src/oneDPL/commit/e625bf013767577392c3b3e28fa1d8e46112f3fb

Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: Fixed.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1041]: Fixed

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1233]: Let's resolve this point a little bit later.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1154]: Probably we have one more option here: to replace this `assert` to `static_assert` and check `sizeof(_Type)`.
If problem really exist I think that there is no reasons to land this PR due it's about only 10% perf profit.

Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1135]: @akukanov what do you think about this? I mean the case described by @danhoeflinger at https://github.com/oneapi-src/oneDPL/pull/1650#discussion_r1658876487 ?

Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1135]: Probably this point has been already resolved early: in the structures `__parallel_find_forward_tag` and `__parallel_find_backward_tag` the `_AtomicType` has been defined as below:
```C++
// FPGA devices don't support 64-bit atomics
#if _ONEDPL_FPGA_DEVICE
    using _AtomicType = uint32_t;
#else
    using _AtomicType = oneapi::dpl::__internal::__difference_t<_RangeType>;
#endif
```

**Review by julianmi — APPROVED** [review]
LGTM. I agree with @danhoeflinger that we should extend this in a separate PR to support 64-bit types. Please wait for his approval before merging.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (danhoeflinger) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1014]: Should this use the same Atomic type as in `__parallel_find_[backward,forward]_tag`?

Follow ups:  
1) Does `backward` have a reason to be signed int rather than unsigned?
2) Do we check that `n` is less than max (index type)?  (perhaps we have other problems with `n` > 2 Billion)

**Review by danhoeflinger — APPROVED** [review]
Basically looks good to me, other than some questions about atomics and making consistent, which can be put in another PR if we don't want to address them here.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1014]: About "Should this use the same Atomic type as in __parallel_find_[backward,forward]_tag?"
 - it's simple not required because we save the fact that we found something, not real position. From this point of view it may be `bool` type too. But I assume that we haven't any additional perf proof from that.

Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1014]: Yes, ok that is right for the `or` case.  However, I think it cant be bool because atomics must be at least 32bits.  

I think my follow ups still make sense for `__parallel_find_[backward,forward]_tag` though.  
Should we align these with each other?

**Pull request merged**
